relay-sampling/src/evaluation.rs

```diff
@@ -24,26 +30,148 @@ fn pseudo_random_from_uuid(id: Uuid) -> f64 {
+}
+
+impl ReservoirEvaluator {
+    /// Constructor for [`ReservoirEvaluator`].
+    pub fn new(counters: ReservoirCounters) -> Self {
+        Self {
+            counters,
+            #[cfg(feature = "redis")]
+            org_id: None,
+            #[cfg(feature = "redis")]
+            redis_pool: None,
+        }
+    }
+
+    /// Sets the Redis pool and organiation ID for the [`ReservoirEvaluator`].
+    ///
+    /// These values are needed to synchronize with Redis.
+    #[cfg(feature = "redis")]
+    pub fn set_redis(mut self, org_id: u64, redis_pool: Arc<RedisPool>) -> Self {
+        self.org_id = Some(org_id);
+        self.redis_pool = Some(redis_pool);
+        self
+    }
+
+    #[allow(unused_variables)]
+    fn redis_count(&self, rule: RuleId, rule_expiry: Option<&DateTime<Utc>>) -> Option<i64> {
+        #[cfg(feature = "redis")]
+        if let (Some(pool), Some(org_id)) = (self.redis_pool.as_ref(), self.org_id) {
+            let key = ReservoirRuleKey::new(org_id, rule);
+
+            let mut redis_client = pool.client().ok()?;
+            let mut redis_connection = redis_client.connection().ok()?;
+
+            if redis_sampling::set_redis_expiry(&mut redis_connection, &key, rule_expiry).is_err() {
+                relay_log::error!("failed to set redis reservoir rule expiry");
+            }
+
+            match redis_sampling::increment_redis_reservoir_count(&mut redis_connection, &key) {
+                Ok(redis_val) => return Some(redis_val),
+                Err(e) => relay_log::error!("failed to increment redis reservoir count: {}", e),
+            }
+        }
+
+        None
+    }
+
+    /// Gets the local count of a reservoir rule. Increments the count if limit isnt reached.
+    fn local_count(&self, rule: RuleId, limit: i64) -> Option<i64> {
+        let mut map_guard = self.counters.lock().ok()?;
+
+        let counter_value = map_guard.entry(rule).or_insert(0);
+        if *counter_value < limit {
+            *counter_value += 1;
+        }
+
+        Some(*counter_value)
+    }
+
+    fn update_counter(&self, rule: RuleId, new_value: i64) {
+        let Ok(mut map_guard) = self.counters.lock() else {
+            relay_log::error!("failed to lock reservoir counter mutex");
+            return;
+        };
+
+        match map_guard.get_mut(&rule) {
+            Some(value) => *value = new_value,
+            // Logging an error because at this point the value should definitively be here.
+            None => relay_log::error!("failed to retrieve counter entry"),
+        }
+    }
+
+    /// Evaluates a reservoir rule, returning `true` if it should be sampled.
+    pub fn evaluate(&self, rule: RuleId, limit: i64, rule_expiry: Option<&DateTime<Utc>>) -> bool {
+        let Some(incremented_local_count) = self.local_count(rule, limit) else {
+            relay_log::error!("failed to read local reservoir count");
+            return false;
+        };
+
+        if incremented_local_count >= limit {
```

Comment:
There's an edge case you could handle better: Assume a limit _N_. Counter is _N - 1_. After running `local_count`, the counter is _N_, and you do not increment Redis, which you should, however.

A fixed version of this would:
- Check the local count without incrementing, bail if _count >= N_.
- Increment redis
- Get the new count
- Return _count <= N_

Reply:
Thank you! I should have been more attentive. 

about your solution, if we check local count without incrementing, then we would have to access the lock once more in the end of the function to update it in case of no redis update.

I changed it now to address the edge case without any additional locking.